### PR TITLE
axiddr3, dma, vsim, generate_constraint.py

### PR DESCRIPTION
**Pull request by hanw**
@jameyhicks These changes has been sitting on our tree for a little while. May be time to merge in.

It contains a fix in 	4c7bfdb from @kslee for a corner case in DMA. 

One piece I am not sure about is the change in 7ebdbd2 to Makefile.connectal.build, I needed it to compile a cpp file to simulate associative memory (using unordered_map in c++). 

/// This is obsolete /// Also, another change in 7ebdbd2 to allow me to simulate xilinx 10G mac, but the change seems like a hack now. I can remove the change if you don't like it.

### CODE REVIEW

**Comment by kslee** [comment]
4c7bfdb
Altera Avalon adds some redundant data to Mem completion packets. As a result, wordCount should be the length of the data following the completion packet, although the completion packet also contains the first 4 bytes of it.

File: scripts/Makefile.connectal.build [hunk 1]
```diff
@@ -325,16 +325,16 @@ VSIM_OBJECTS = $(foreach c, $(VSIM_CFILES), vsim.dir/xsc/$(c).lnx64.o ) $(foreac
```
Comment (jameyhicks) [inline scripts/Makefile.connectal.build]: The source file here is .c, so it should be gcc.

File: scripts/Makefile.connectal.build [hunk 2]
```diff
@@ -325,16 +325,16 @@ VSIM_OBJECTS = $(foreach c, $(VSIM_CFILES), vsim.dir/xsc/$(c).lnx64.o ) $(foreac
 
 vsim.dir/xsc/%.lnx64.o: $(CONNECTALDIR)/cpp/%.c
 	mkdir -p vsim.dir/xsc
-	gcc -c $(VSIM_CFLAGS) -O1 -o vsim.dir/xsc/$(*).lnx64.o $<
+	g++ -c $(VSIM_CFLAGS) -O1 -o vsim.dir/xsc/$(*).lnx64.o $<
 vsim.dir/xsc/%.lnx64.o: $(CONNECTALDIR)/cpp/%.cpp
```
Comment (jameyhicks) [inline scripts/Makefile.connectal.build]: but yes, this one should be g++. How did I miss that?

File: scripts/Makefile.connectal.build [hunk 3]
```diff
@@ -388,7 +388,7 @@ obj/Makefile: $(OBJMAKEFILE_DEP) syntax.timestamp
 
 $(DTOP)/bin/libconnectal-sim.so: $(SIM_CXX_COMMON) $(SIM_CXX_PROJECT) prepare_bin_target
 	$(Q)mkdir -p $(DTOP)/bin
-	g++ -O -g -I$(DTOP)/jni -shared -fpic $(SIM_CXXFLAGS) -g -o $(DTOP)/bin/libconnectal-sim.so $(SIM_CXX_LOCAL) $(SIM_CXX_PROJECT)
+	g++ -O -g -I$(DTOP)/jni -shared -fpic -std=c++11 $(SIM_CXXFLAGS) -g -o $(DTOP)/bin/libconnectal-sim.so $(SIM_CXX_LOCAL) $(SIM_CXX_PROJECT)
```
Comment (jameyhicks) [inline scripts/Makefile.connectal.build]: why is that flag needed here? If it is project specific, is there some way we can pass it via CONNECTALFLAGS?

File: scripts/connectal-synth-axiddr3.tcl [hunk 4]
```diff
@@ -4,4 +4,4 @@ source "$connectaldir/scripts/connectal-synth-ip.tcl"
 set prj_boardname $boardname
 if [string match "*g2" $boardname] {set prj_boardname [string trimright $boardname "g2"]}
```
Comment (jameyhicks) [inline scripts/connectal-synth-axiddr3.tcl]: this number should be based on vivado version number or else we accept any version

File: bsv/MemWriteEngine.bsv [hunk 5]
```diff
@@ -105,8 +105,14 @@ module mkMemWriteChannel#(Integer bufferSizeBytes, Integer channelNumber,
       if (clientBursts) begin
 	 load_in_progress <= True;
 	 let cmd = clientStart;
-	 let cond0 <- clientAvail.maybeDecrement(unpack(extend(cmd.burstLen>>beat_shift)));
 	 let cond1 = cmd.len <= extend(cmd.burstLen);
+         Bool cond0 = False;
```
Comment (jameyhicks) [inline bsv/MemWriteEngine.bsv]: oh, this is a problem. maybe. I'm worried about the critical path. But better to have a long critical path than incorrect operation.

File: scripts/generate-constraints.py [hunk 6]
```diff
@@ -83,6 +83,9 @@ def newArgparser():
             used = []
             boardPinInfo = {}
             pinName = ''
+            pinFound = len([x for x in projectPinInfo.keys() if x in bindings.keys()])
```
Comment (jameyhicks) [inline scripts/generate-constraints.py]: This script has gotten a bit unruly, but the idea is that common pins can be specified in the boardinfo file and other ones can be fully specified in the project pinout.json. This change would prevent a pin from being fully specified in the pinout.json file. Please revert this change to generate-constraints.py

Comment (hanw) [inline scripts/generate-constraints.py]: removed.

File: scripts/Makefile.connectal.build (hunk 3 above)
Comment (hanw) [inline scripts/Makefile.connectal.build]: removed.

File: scripts/Makefile.connectal.build (hunk 1 above)
Comment (hanw) [inline scripts/Makefile.connectal.build]: fixed.

**Comment by hanw** [comment]
Can we merge now?
